// @chipdeals/payments/lib/utils/commons.js
"use strict";
class Commons {
    static isEmpty(data) {
        if (/(number)|(boolean)/.test(typeof data)) {
            return false;
        }
        if (!data) {
            return true;
        }
        if (Array.isArray(data)) {
            return data.length === 0;
        }
        const dataProperties = Object.values(data);
        return dataProperties && dataProperties.length === 0;
    }
    static consoleDate() {
        const date = new Date();
        let displayDate = '';
        displayDate += date.getUTCFullYear() + '-';
        displayDate += ('0' + (date.getUTCMonth() + 1)).replace(/.+(.{2})$/, '$1') + '-';
        displayDate += ('0' + date.getUTCDate()).replace(/.+(.{2})$/, '$1') + ' ';
        displayDate += ('0' + date.getUTCHours()).replace(/.+(.{2})$/, '$1') + ':';
        displayDate += ('0' + date.getUTCMinutes()).replace(/.+(.{2})$/, '$1') + ':';
        displayDate += ('0' + date.getUTCSeconds()).replace(/.+(.{2})$/, '$1') + ':';
        displayDate += ('00' + date.getUTCMilliseconds()).replace(/.+(.{3})$/, '$1');
        return displayDate;
    }
    static serverBaseUrl() {
        return 'https://apis.chipdeals.me/payment';
    }
    static cardHostToHostServerBaseUrl() {
        return 'https://apis.chipdeals.me/card-host-to-host';
    }
    static isTestApiKey(apiKey) {
        const testApiKeyRegex = /^test/;
        return testApiKeyRegex.test(apiKey);
    }
}
module.exports = Commons;
