// @chipdeals/payments/lib/utils/requestUtils.js
"use strict";
// @ts-check
const Axios = require("axios");
const AxiosRetry = require("axios-retry");
const Commons = require("./commons");
const https = require("https");
const axios = Axios.default;
const axiosRetry = AxiosRetry.default;
axiosRetry(axios, {
    retries: 100,
    retryDelay: axiosRetry.exponentialDelay,
    retryCondition: (error) => !error.response,
    shouldResetTimeout: true,
});
class RequestUtils {
    static sendRequestTo(url, data, headers, method, getStatus = false) {
        return request(method, url, headers, null, data, null, undefined, getStatus);
    }
}
async function request(method, url, headers = null, auth = null, data = null, responseType = null, retryNumber = 5, returnStatus = false) {
    const agent = new https.Agent({
        rejectUnauthorized: false,
    });
    const requestConfig = {
        url,
        method,
        timeout: 60000,
        'axios-retry': {
            retries: retryNumber,
        },
        httpsAgent: agent,
    };
    if (!Commons.isEmpty(headers))
        requestConfig.headers = headers;
    if (!Commons.isEmpty(auth))
        requestConfig.auth = auth;
    if (!Commons.isEmpty(data))
        requestConfig.data = data;
    if (!Commons.isEmpty(responseType))
        requestConfig.responseType = responseType;
    const response = await axios.request(requestConfig).catch((error) => error.response || error);
    if (returnStatus) {
        return {
            status: { code: response.status },
            response: response.data || response,
        };
    }
    return response.data || response;
}
module.exports = RequestUtils;
