// @chipdeals/payments/lib/collection/collectRequest.js
"use strict";
const CollectionExecution = require("./collectionExecution");
class CollectRequest {
    constructor(apiKey, enableLogs) {
        this._enableLogs = false;
        this._apiKey = '';
        this._amount = 0;
        this._currency = '';
        this._fee = 0;
        this._merchantSupportFee = true;
        this._phoneNumber = '';
        this._firstName = 'anonymous';
        this._lastName = 'anonymous';
        this._webhookUrl = '';
        this._otp = '';
        this._apiKey = apiKey;
        this._enableLogs = enableLogs;
    }
    amount(amount) {
        this._amount = amount;
        return this;
    }
    currency(currency) {
        this._currency = currency;
        return this;
    }
    setFee(fee, userSupportAllFees = false) {
        this._fee = fee;
        this._merchantSupportFee = !userSupportAllFees;
        return this;
    }
    from(phoneNumber) {
        this._phoneNumber = phoneNumber;
        return this;
    }
    firstName(firstName) {
        this._firstName = firstName;
        return this;
    }
    lastName(lastName) {
        this._lastName = lastName;
        return this;
    }
    webhook(webhookUrl) {
        this._webhookUrl = webhookUrl;
        return this;
    }
    useOtp(otp) {
        this._otp = otp;
        return this;
    }
    create(onCreateCallback = () => {
        return null;
    }) {
        const collectionCallback = new CollectionExecution({
            apiKey: this._apiKey,
            amount: this._amount,
            currency: this._currency,
            fee: this._fee,
            merchantSupportFee: this._merchantSupportFee,
            phoneNumber: this._phoneNumber,
            firstName: this._firstName,
            lastName: this._lastName,
            webhookUrl: this._webhookUrl,
            otp: this._otp,
        }, onCreateCallback, this._enableLogs);
        return collectionCallback;
    }
}
module.exports = CollectRequest;
