// @chipdeals/payments/lib/disbursement/disbursementExecution.js
"use strict";
const chalk = require("chalk");
const getTransactionStatus = require("../use-cases/getTransactionStatus");
const sendDisbursementRequest = require("../use-cases/sendDisbursementRequest");
const Commons = require("../utils/commons");
class CollectionExecution {
    constructor({ apiKey, amount, currency, phoneNumber, webhookUrl, }, onCreateCallback = () => {
        return null;
    }, enableLogs) {
        this._enableLogs = false;
        this._apiKey = '';
        this._amount = 0;
        this._currency = '';
        this._phoneNumber = '';
        this._webhookUrl = '';
        this._apiKey = apiKey;
        this._amount = amount;
        this._currency = currency;
        this._phoneNumber = phoneNumber;
        this._webhookUrl = webhookUrl;
        this._enableLogs = enableLogs;
        this._onCreateCallback = onCreateCallback;
        if (!this._amount || !this._currency || !this._phoneNumber) {
            this._transactionData = {
                reference: '',
                recipientPhoneNumber: '',
                recipientOperator: '',
                currency: '',
                amount: 0,
                status: 'error',
                statusMessage: 'an error occured, transaction',
                statusMessageCode: 555,
                startTimestampInSecond: 0,
                endTimestampInSecond: 0,
            };
            if (!this._amount) {
                this._transactionData.statusMessage = 'You must add the amount to collect with method .amount';
                this._transactionData.statusMessageCode = 400;
                this._transactionData.status = 'error';
            }
            if (!this._currency) {
                this._transactionData.statusMessage = 'You must add the currency of disbursement with method .currency';
                this._transactionData.statusMessageCode = 400;
                this._transactionData.status = 'error';
            }
            if (!this._phoneNumber) {
                this._transactionData.statusMessage = 'You must add the phoneNumber with method .from';
                this._transactionData.statusMessageCode = 400;
                this._transactionData.status = 'error';
            }
            this._onErrorCallback(this._transactionData);
        }
        this.startRequest();
    }
    _onCreateCallback(transactionReference) {
        return null;
    }
    _onSuccessCallback(transactionData) {
        logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
    }
    _onStatusChangedCallback(transactionData) {
        logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
    }
    _onErrorCallback(transactionData) {
        logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
    }
    onSuccess(onSuccessCallback) {
        var _a;
        this._onSuccessCallback = (transactionData) => {
            logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
            return onSuccessCallback(transactionData);
        };
        if (((_a = this._transactionData) === null || _a === void 0 ? void 0 : _a.status) === 'success') {
            this._onSuccessCallback(this._transactionData);
        }
        return this;
    }
    onStatusChanged(onStatusChangedCallback) {
        this._onStatusChangedCallback = (transactionData) => {
            logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
            return onStatusChangedCallback(transactionData);
        };
        return this;
    }
    onError(onErrorCallback) {
        var _a;
        this._onErrorCallback = (transactionData) => {
            logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
            return onErrorCallback(transactionData);
        };
        if (((_a = this._transactionData) === null || _a === void 0 ? void 0 : _a.status) === 'error') {
            this._onErrorCallback(this._transactionData);
        }
        return this;
    }
    async startRequest() {
        console.log(chalk.gray(Commons.consoleDate()) + ' Disbursement request');
        const disbursementResponse = await sendDisbursementRequest({
            apiKey: this._apiKey,
            amount: this._amount,
            currency: this._currency,
            phoneNumber: this._phoneNumber,
            webhookUrl: this._webhookUrl,
        });
        const transactionData = Object.assign(Object.assign({ reference: '', recipientPhoneNumber: '', recipientOperator: '', currency: '', amount: 0, status: 'error', statusMessage: disbursementResponse.message || 'an error occured', startTimestampInSecond: 0, endTimestampInSecond: 0 }, disbursementResponse.deposit), { statusMessageCode: disbursementResponse.deposit.statusMessageCode || 556 });
        this._transactionData = transactionData;
        if (!Commons.isEmpty(transactionData.reference)) {
            this._onCreateCallback(transactionData.reference);
        }
        if (!disbursementResponse.success) {
            this._onErrorCallback(transactionData);
            return;
        }
        this._onStatusChangedCallback(transactionData);
        await new Promise((resolve) => {
            setTimeout(resolve, 2000);
        });
        this.checkTransactionStatus();
    }
    async checkTransactionStatus() {
        if (!this._transactionData) {
            this._transactionData = {
                reference: '',
                recipientPhoneNumber: '',
                recipientOperator: '',
                currency: '',
                amount: 0,
                status: 'error',
                statusMessage: 'an error occured, transaction',
                statusMessageCode: 557,
                startTimestampInSecond: 0,
                endTimestampInSecond: 0,
            };
            this._onErrorCallback(this._transactionData);
            return;
        }
        const statusResponse = await getTransactionStatus(this._apiKey, this._transactionData.reference);
        const transactionData = Object.assign(Object.assign(Object.assign(Object.assign({}, this._transactionData), { status: 'error', statusMessage: statusResponse.message || 'an error occured', startTimestampInSecond: 0, endTimestampInSecond: 0 }), statusResponse.transaction), { statusMessageCode: statusResponse.transaction.statusMessageCode || 558 });
        let recheckStatus = true;
        if (!statusResponse.success) {
            this._onErrorCallback(transactionData);
            recheckStatus = false;
        }
        else if (/error/.test(transactionData.status)) {
            this._onErrorCallback(transactionData);
            recheckStatus = false;
        }
        else if (/success/.test(transactionData.status)) {
            this._onSuccessCallback(transactionData);
            recheckStatus = false;
        }
        else if (transactionData.statusMessageCode !== this._transactionData.statusMessageCode) {
            this._onStatusChangedCallback(transactionData);
        }
        this._transactionData = transactionData;
        if (recheckStatus) {
            await new Promise((resolve) => {
                setTimeout(resolve, 2000);
            });
            this.checkTransactionStatus();
        }
    }
}
function logTransactionStatus(transactionData, isTestRequest, enableLogs) {
    if (!enableLogs)
        return;
    console.log(chalk.gray(Commons.consoleDate()) +
        ' ' +
        (isTestRequest ? chalk.inverse('TEST') + ' ' : '') +
        '[' +
        chalk.blue(transactionData.statusMessageCode) +
        '] ' +
        transactionData.statusMessage);
    if (transactionData.status === 'success') {
        console.log(chalk.gray(Commons.consoleDate()) +
            ' ' +
            (isTestRequest ? chalk.inverse('TEST') + ' ' : '') +
            '[' +
            chalk.green('OK') +
            '] ' +
            transactionData.amount +
            ' ' +
            transactionData.currency +
            ' sent to customer');
    }
    if (transactionData.status === 'error') {
        console.log(chalk.gray(Commons.consoleDate()) +
            ' ' +
            (isTestRequest ? chalk.inverse('TEST') + ' ' : '') +
            '[' +
            chalk.red('ERROR') +
            '] transaction aborted');
    }
}
module.exports = CollectionExecution;
