// @chipdeals/payments/lib/disbursement/disbursementRequest.js
"use strict";
const DisbursementExecution = require("./disbursementExecution");
class DisbursementRequest {
    constructor(apiKey, enableLogs) {
        this._enableLogs = false;
        this._apiKey = '';
        this._amount = 0;
        this._currency = '';
        this._phoneNumber = '';
        this._webhookUrl = '';
        this._apiKey = apiKey;
        this._enableLogs = enableLogs;
    }
    amount(amount) {
        this._amount = amount;
        return this;
    }
    currency(currency) {
        this._currency = currency;
        return this;
    }
    to(phoneNumber) {
        this._phoneNumber = phoneNumber;
        return this;
    }
    webhook(webhookUrl) {
        this._webhookUrl = webhookUrl;
        return this;
    }
    create(onCreateCallback = () => {
        return null;
    }) {
        const collectionCallback = new DisbursementExecution({
            apiKey: this._apiKey,
            amount: this._amount,
            currency: this._currency,
            phoneNumber: this._phoneNumber,
            webhookUrl: this._webhookUrl,
        }, onCreateCallback, this._enableLogs);
        return collectionCallback;
    }
}
module.exports = DisbursementRequest;
