// @chipdeals/payments/lib/use-cases/sendCardCollectionRequest.js
"use strict";
const Commons = require("../utils/commons");
const RequestUtils = require("../utils/requestUtils");
async function sendCardCollectionRequest({ apiKey, encApiKey, amount, currency, fee, merchantSupportFee, phoneNumber, firstName, lastName, title, email, address, locality, district, postalCode, noCard, cvv, expiry, webhookUrl, otp, }) {
    const requestData = {
        apiKey: encApiKey,
        senderFirstName: firstName,
        senderLastName: lastName,
        senderPhoneNumber: phoneNumber,
        title,
        email,
        address,
        locality,
        district,
        postalCode,
        noCard,
        cvv,
        expiry,
        fee,
        merchantSupportFee,
        currency,
        amount,
        webhookUrl,
        otp,
    };
    const collectionResponse = await RequestUtils.sendRequestTo(Commons.cardHostToHostServerBaseUrl() + '/launchpayment?apikey=' + apiKey, requestData, null, 'post', true);
    if (collectionResponse.response.error && collectionResponse.response.error.msg) {
        return {
            success: false,
            message: collectionResponse.response.error.msg,
            payment: { statusMessageCode: collectionResponse.status.code },
        };
    }
    if (collectionResponse.response.error) {
        return {
            success: false,
            message: 'error',
            payment: { statusMessageCode: collectionResponse.status.code },
        };
    }
    if (!/^2../.test(collectionResponse.status.code)) {
        return Object.assign(Object.assign({ success: false, message: 'error' }, collectionResponse.response), { payment: Object.assign(Object.assign({}, (collectionResponse.response.transaction || {})), { statusMessageCode: collectionResponse.status.code }) });
    }
    const response = Object.assign(Object.assign({}, collectionResponse.response), { payment: Object.assign({ statusMessageCode: collectionResponse.status.code }, (collectionResponse.response.transaction || {})) });
    delete response.transaction.successRedirectionUrl;
    delete response.transaction.errorRedirectionUrl;
    delete response.payment.successRedirectionUrl;
    delete response.payment.errorRedirectionUrl;
    return response;
}
module.exports = sendCardCollectionRequest;
