// @chipdeals/payments/lib/use-cases/sendCollectionRequest.js
"use strict";
const Commons = require("../utils/commons");
const RequestUtils = require("../utils/requestUtils");
async function sendCollectionRequest({ apiKey, amount, currency, fee, merchantSupportFee, phoneNumber, firstName, lastName, webhookUrl, otp, }) {
    const requestData = {
        senderFirstName: firstName,
        senderLastName: lastName,
        senderPhoneNumber: phoneNumber,
        fee,
        merchantSupportFee,
        currency,
        amount,
        webhookUrl,
        otp,
    };
    const collectionResponse = await RequestUtils.sendRequestTo(Commons.serverBaseUrl() + '/requestPayment?apikey=' + apiKey, requestData, null, 'post', true);
    if (collectionResponse.response.error && collectionResponse.response.error.msg) {
        return {
            success: false,
            message: collectionResponse.response.error.msg,
            payment: { statusMessageCode: collectionResponse.status.code },
        };
    }
    if (collectionResponse.response.error) {
        return {
            success: false,
            message: 'error',
            payment: { statusMessageCode: collectionResponse.status.code },
        };
    }
    if (!/^2../.test(collectionResponse.status.code)) {
        return Object.assign(Object.assign({ success: false, message: 'error' }, collectionResponse.response), { payment: Object.assign(Object.assign({}, (collectionResponse.response.payment || {})), { statusMessageCode: collectionResponse.status.code }) });
    }
    return collectionResponse.response;
}
module.exports = sendCollectionRequest;
