// @chipdeals/payments/lib/chipdealsPayment.js
"use strict";
const CollectionRequest = require("./collection/collectRequest");
const CardCollectionRequest = require("./cardCollection/collectRequest");
const DisbursementRequest = require("./disbursement/disbursementRequest");
const getBalance = require("./use-cases/getBalance");
const getTransactionStatus = require("./use-cases/getTransactionStatus");
const Commons = require("./utils/commons");
class ChipdealsPayment {
    constructor() {
        this._apiKey = '';
        this._secretKey = '';
        this._enableLogs = false;
        this._apiKeyError = 'apiKey not found, add apiKey with momo.setApiKey method';
        this._collectionRequests = [];
        this._cardCollectionRequests = [];
        this._disbursementRequests = [];
    }
    setApiKey(apiKey) {
        this._apiKey = apiKey;
        this._enableLogs = Commons.isTestApiKey(this._apiKey);
        return this;
    }
    setSecretKey(secretKey) {
        this._secretKey = secretKey;
        return this;
    }
    logs(enableLogs = Commons.isTestApiKey(this._apiKey) || true) {
        this._enableLogs = enableLogs;
        return this;
    }
    collect() {
        if (Commons.isEmpty(this._apiKey))
            throw new Error(this._apiKeyError);
        const collection = new CollectionRequest(this._apiKey, this._enableLogs);
        this._collectionRequests.push(collection);
        return collection;
    }
    collectWithCard() {
        if (Commons.isEmpty(this._apiKey))
            throw new Error(this._apiKeyError);
        const collection = new CardCollectionRequest(this._apiKey, this._secretKey, this._enableLogs);
        this._cardCollectionRequests.push(collection);
        return collection;
    }
    deposit() {
        if (Commons.isEmpty(this._apiKey))
            throw new Error(this._apiKeyError);
        const disbursement = new DisbursementRequest(this._apiKey, this._enableLogs);
        this._disbursementRequests.push(disbursement);
        return disbursement;
    }
    async balance() {
        if (Commons.isEmpty(this._apiKey))
            throw new Error(this._apiKeyError);
        const balanceResponse = await getBalance(this._apiKey);
        if (!balanceResponse.success) {
            throw new Error(balanceResponse.message);
        }
        return balanceResponse.balances;
    }
    async status(reference) {
        if (Commons.isEmpty(this._apiKey))
            throw new Error(this._apiKeyError);
        const statusResponse = await getTransactionStatus(this._apiKey, reference);
        const response = Object.assign({ type: Commons.isEmpty(statusResponse.transaction.recipientPhoneNumber) ? 'collection' : 'deposit' }, statusResponse.transaction);
        delete response.checkoutUrl;
        delete response.successRedirectionUrl;
        delete response.errorRedirectionUrl;
        return response;
    }
}
module.exports = ChipdealsPayment;
