// @chipdeals/payments/lib/cardCollection/collectionExecution.js
"use strict";
const getTransactionStatus = require("../use-cases/getTransactionStatus");
const sendCardCollectionRequest = require("../use-cases/sendCardCollectionRequest");
const Commons = require("../utils/commons");
const chalk = require("chalk");
const CryptoJS = require("crypto-js");
class CollectionExecution {
    constructor({ apiKey, secretKey, amount, currency, fee, merchantSupportFee, phoneNumber, firstName, lastName, title, email, address, locality, district, postalCode, noCard, cvv, expiry, webhookUrl, otp, }, onCreateCallback = () => {
        return null;
    }, enableLogs) {
        this._enableLogs = false;
        this._apiKey = '';
        this._secretKey = '';
        this._amount = 0;
        this._currency = '';
        this._fee = 0;
        this._merchantSupportFee = true;
        this._phoneNumber = '';
        this._firstName = '';
        this._lastName = '';
        this._title = '';
        this._email = '';
        this._address = '';
        this._locality = '';
        this._district = '';
        this._postalCode = '';
        this._noCard = '';
        this._cvv = '';
        this._expiry = '';
        this._webhookUrl = '';
        this._otp = '';
        this._apiKey = apiKey;
        this._secretKey = secretKey;
        this._amount = amount;
        this._currency = currency;
        this._phoneNumber = phoneNumber;
        this._fee = fee;
        this._merchantSupportFee = merchantSupportFee;
        this._firstName = firstName;
        this._lastName = lastName;
        this._title = title;
        this._email = email;
        this._address = address;
        this._locality = locality;
        this._district = district;
        this._postalCode = postalCode;
        this._webhookUrl = webhookUrl;
        this._noCard = noCard;
        this._cvv = cvv;
        this._expiry = expiry;
        this._otp = otp;
        this._enableLogs = enableLogs;
        this._onCreateCallback = onCreateCallback;
        if (!this._amount || !this._currency || !this._phoneNumber) {
            this._transactionData = {
                reference: '',
                senderPhoneNumber: '',
                senderOperator: '',
                senderFirstName: '',
                senderLastName: '',
                currency: '',
                amount: 0,
                status: 'error',
                statusMessage: 'an error occured, transaction',
                statusMessageCode: 551,
                startTimestampInSecond: 0,
                endTimestampInSecond: 0,
            };
            if (!this._amount) {
                this._transactionData.statusMessage = 'You must add the amount to collect with method .amount';
                this._transactionData.statusMessageCode = 400;
                this._transactionData.status = 'error';
            }
            if (!this._currency) {
                this._transactionData.statusMessage = 'You must add the currency of collection with method .currency';
                this._transactionData.statusMessageCode = 400;
                this._transactionData.status = 'error';
            }
            if (!this._phoneNumber) {
                this._transactionData.statusMessage = 'You must add the phoneNumber with method .from';
                this._transactionData.statusMessageCode = 400;
                this._transactionData.status = 'error';
            }
            this._onErrorCallback(this._transactionData);
        }
        this.startRequest();
    }
    _onCreateCallback(transactionReference) {
        return null;
    }
    _onSuccessCallback(transactionData) {
        logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
    }
    _onStatusChangedCallback(transactionData) {
        logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
    }
    _onErrorCallback(transactionData) {
        logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
    }
    onSuccess(onSuccessCallback) {
        var _a;
        this._onSuccessCallback = (transactionData) => {
            logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
            return onSuccessCallback(transactionData);
        };
        if (((_a = this._transactionData) === null || _a === void 0 ? void 0 : _a.status) === 'success') {
            this._onSuccessCallback(this._transactionData);
        }
        return this;
    }
    onStatusChanged(onStatusChangedCallback) {
        this._onStatusChangedCallback = (transactionData) => {
            logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
            return onStatusChangedCallback(transactionData);
        };
        return this;
    }
    onError(onErrorCallback) {
        var _a;
        this._onErrorCallback = (transactionData) => {
            logTransactionStatus(transactionData, Commons.isTestApiKey(this._apiKey), this._enableLogs);
            return onErrorCallback(transactionData);
        };
        if (((_a = this._transactionData) === null || _a === void 0 ? void 0 : _a.status) === 'error') {
            this._onErrorCallback(this._transactionData);
        }
        return this;
    }
    async startRequest() {
        if (this._enableLogs)
            console.log(chalk.gray(Commons.consoleDate()) + ' Collection request');
        const { encNoCard, encExpiry, encCvv, encApiKey } = encryptData(this._noCard, this._cvv, this._expiry, this._apiKey, this._secretKey);
        const collectionResponse = await sendCardCollectionRequest({
            apiKey: this._apiKey,
            encApiKey,
            amount: this._amount,
            currency: this._currency,
            fee: this._fee,
            merchantSupportFee: this._merchantSupportFee,
            phoneNumber: this._phoneNumber,
            firstName: this._firstName,
            lastName: this._lastName,
            title: this._title,
            email: this._email,
            address: this._address,
            locality: this._locality,
            district: this._district,
            postalCode: this._postalCode,
            noCard: encNoCard,
            cvv: encCvv,
            expiry: encExpiry,
            webhookUrl: this._webhookUrl,
            otp: this._otp,
        });
        const transactionData = Object.assign(Object.assign({ reference: '', senderPhoneNumber: '', senderOperator: '', senderFirstName: '', senderLastName: '', currency: '', amount: 0, status: 'error', statusMessage: collectionResponse.message || 'an error occured', startTimestampInSecond: 0, endTimestampInSecond: 0 }, collectionResponse.payment), { statusMessageCode: collectionResponse.payment.statusMessageCode || 552 });
        this._transactionData = transactionData;
        if (!Commons.isEmpty(transactionData.reference)) {
            this._onCreateCallback(transactionData.reference);
        }
        if (!collectionResponse.success) {
            this._onErrorCallback(transactionData);
            return;
        }
        this._onStatusChangedCallback(transactionData);
        await new Promise((resolve) => {
            setTimeout(resolve, 2000);
        });
        this.checkTransactionStatus();
    }
    async checkTransactionStatus() {
        if (!this._transactionData) {
            this._transactionData = {
                reference: '',
                senderPhoneNumber: '',
                senderOperator: '',
                senderFirstName: '',
                senderLastName: '',
                currency: '',
                amount: 0,
                status: 'error',
                statusMessage: 'an error occured, transaction',
                statusMessageCode: 553,
                startTimestampInSecond: 0,
                endTimestampInSecond: 0,
            };
            this._onErrorCallback(this._transactionData);
            return;
        }
        const statusResponse = await getTransactionStatus(this._apiKey, this._transactionData.reference);
        const transactionData = Object.assign(Object.assign(Object.assign(Object.assign({}, this._transactionData), { status: 'error', statusMessage: statusResponse.message || 'an error occured', startTimestampInSecond: 0, endTimestampInSecond: 0 }), statusResponse.transaction), { statusMessageCode: statusResponse.transaction.statusMessageCode || 554 });
        let recheckStatus = true;
        if (!statusResponse.success) {
            this._onErrorCallback(transactionData);
            recheckStatus = false;
        }
        else if (/error/.test(transactionData.status)) {
            this._onErrorCallback(transactionData);
            recheckStatus = false;
        }
        else if (/success/.test(transactionData.status)) {
            this._onSuccessCallback(transactionData);
            recheckStatus = false;
        }
        else if (transactionData.statusMessageCode !== this._transactionData.statusMessageCode) {
            this._onStatusChangedCallback(transactionData);
        }
        this._transactionData = transactionData;
        if (recheckStatus) {
            await new Promise((resolve) => {
                setTimeout(resolve, 2000);
            });
            this.checkTransactionStatus();
        }
    }
}
function logTransactionStatus(transactionData, isTestRequest, enableLogs) {
    if (!enableLogs)
        return;
    console.log(chalk.gray(Commons.consoleDate()) +
        ' ' +
        (isTestRequest ? chalk.inverse('TEST') + ' ' : '') +
        '[' +
        chalk.blue(transactionData.statusMessageCode) +
        '] ' +
        transactionData.statusMessage);
    if (transactionData.status === 'success') {
        console.log(chalk.gray(Commons.consoleDate()) +
            ' ' +
            (isTestRequest ? chalk.inverse('TEST') + ' ' : '') +
            '[' +
            chalk.green('OK') +
            '] ' +
            transactionData.amount +
            ' ' +
            transactionData.currency +
            ' received form customer');
    }
    if (transactionData.status === 'error') {
        console.log(chalk.gray(Commons.consoleDate()) +
            ' ' +
            (isTestRequest ? chalk.inverse('TEST') + ' ' : '') +
            '[' +
            chalk.red('ERROR') +
            '] transaction aborted');
    }
}
function encryptData(noCard, cvv, expiry, apiKey, secretKey) {
    if (!secretKey)
        return {
            encNoCard: 'noSecretKey',
            encCvv: 'noSecretKey',
            encExpiry: 'noSecretKey',
            encApiKey: 'noSecretKey',
        };
    const encNoCard = CryptoJS.AES.encrypt(noCard, secretKey).toString();
    const encCvv = CryptoJS.AES.encrypt(cvv, secretKey).toString();
    const encExpiry = CryptoJS.AES.encrypt(expiry, secretKey).toString();
    const encApiKey = CryptoJS.AES.encrypt(apiKey, secretKey).toString();
    return { encNoCard, encCvv, encExpiry, encApiKey };
}
module.exports = CollectionExecution;
