// @chipdeals/payments/lib/cardCollection/collectRequest.js
"use strict";
const CollectionExecution = require("./collectionExecution");
class CollectRequest {
    constructor(apiKey, secretKey, enableLogs) {
        this._enableLogs = false;
        this._apiKey = '';
        this._secretKey = '';
        this._amount = 0;
        this._currency = '';
        this._fee = 0;
        this._merchantSupportFee = true;
        this._phoneNumber = '';
        this._firstName = '';
        this._lastName = '';
        this._title = '';
        this._email = '';
        this._address = '';
        this._locality = '';
        this._district = '';
        this._postalCode = '';
        this._noCard = '';
        this._cvv = '';
        this._expiry = '';
        this._webhookUrl = '';
        this._otp = '';
        this._apiKey = apiKey;
        this._secretKey = secretKey;
        this._enableLogs = enableLogs;
    }
    amount(amount) {
        this._amount = amount;
        return this;
    }
    currency(currency) {
        this._currency = currency;
        return this;
    }
    setFee(fee, userSupportAllFees = false) {
        this._fee = fee;
        this._merchantSupportFee = !userSupportAllFees;
        return this;
    }
    phoneNumber(phoneNumber) {
        this._phoneNumber = phoneNumber;
        return this;
    }
    firstName(firstName) {
        this._firstName = firstName;
        return this;
    }
    lastName(lastName) {
        this._lastName = lastName;
        return this;
    }
    title(title) {
        this._title = title;
        return this;
    }
    email(email) {
        this._email = email;
        return this;
    }
    address(address) {
        this._address = address;
        return this;
    }
    locality(locality) {
        this._locality = locality;
        return this;
    }
    district(district) {
        this._district = district;
        return this;
    }
    postalCode(postalCode) {
        this._postalCode = postalCode;
        return this;
    }
    noCard(noCard) {
        this._noCard = noCard;
        return this;
    }
    cvv(cvv) {
        this._cvv = cvv;
        return this;
    }
    expiry(expiry) {
        this._expiry = expiry;
        return this;
    }
    webhook(webhookUrl) {
        this._webhookUrl = webhookUrl;
        return this;
    }
    // useOtp(otp: string): CollectRequest {
    //   this._otp = otp;
    //   return this;
    // }
    create(onCreateCallback = () => {
        return null;
    }) {
        const collectionCallback = new CollectionExecution({
            apiKey: this._apiKey,
            secretKey: this._secretKey,
            amount: this._amount,
            currency: this._currency,
            fee: this._fee,
            merchantSupportFee: this._merchantSupportFee,
            phoneNumber: this._phoneNumber,
            firstName: this._firstName,
            lastName: this._lastName,
            title: this._title,
            email: this._email,
            address: this._address,
            locality: this._locality,
            district: this._district,
            postalCode: this._postalCode,
            noCard: this._noCard,
            cvv: this._cvv,
            expiry: this._expiry,
            webhookUrl: this._webhookUrl,
            otp: this._otp,
        }, onCreateCallback, this._enableLogs);
        return collectionCallback;
    }
}
module.exports = CollectRequest;
